File: tests/src/org.apache.pekko/pekko-actor_2.13/1.6.0/src/test/scala/org_apache_pekko/pekko_actor_2_13/ClassLoaderObjectInputStreamTest.scala
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_apache_pekko.pekko_actor_2_13

import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.ObjectOutputStream
import java.util.ArrayList

import org.apache.pekko.util.ClassLoaderObjectInputStream
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class ClassLoaderObjectInputStreamTest {
  @Test
  def fallsBackToDefaultResolverWhenConfiguredClassLoaderCannotLoadClass(): Unit = {
    val original: ArrayList[String] = new ArrayList[String]()
    original.add("pekko")

    val rejectedClassName: String = classOf[ArrayList[_]].getName
    val input: ClassLoaderObjectInputStream = new ClassLoaderObjectInputStream(
      new ClassLoaderObjectInputStreamTest.RejectingClassLoader(rejectedClassName),
      new ByteArrayInputStream(serialize(original)))

    try {
      val restored: AnyRef = input.readObject()

      assertThat(restored).isInstanceOf(classOf[ArrayList[_]])
      val restoredList: ArrayList[String] = restored.asInstanceOf[ArrayList[String]]
      assertThat(restoredList).hasSize(1)
      assertThat(restoredList.get(0)).isEqualTo("pekko")
    } finally {
      input.close()
    }
  }

  private def serialize(value: AnyRef): Array[Byte] = {
    val bytes: ByteArrayOutputStream = new ByteArrayOutputStream()
    val output: ObjectOutputStream = new ObjectOutputStream(bytes)
    try {
      output.writeObject(value)
    } finally {
      output.close()
    }
    bytes.toByteArray
  }
}

object ClassLoaderObjectInputStreamTest {
  private final class RejectingClassLoader(rejectedClassName: String)
      extends ClassLoader(ClassLoader.getSystemClassLoader) {
    override def loadClass(name: String, resolve: Boolean): Class[_] = {
      if (name == rejectedClassName) {
        throw new ClassNotFoundException(name)
      }
      super.loadClass(name, resolve)
    }
  }
}

File: tests/src/org.testcontainers/testcontainers/1.19.8/src/test/java/org_testcontainers/testcontainers/ClassLoaderObjectInputStreamTest.java
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_testcontainers.testcontainers;

import org.junit.jupiter.api.Test;
import org.testcontainers.shaded.org.apache.commons.io.input.ClassLoaderObjectInputStream;
import org.testcontainers.shaded.org.apache.commons.lang3.SerializationUtils;
import org.testcontainers.shaded.com.google.common.reflect.Reflection;

import java.io.ByteArrayInputStream;
import java.io.Serializable;
import java.lang.reflect.InvocationHandler;
import java.util.ArrayList;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

public class ClassLoaderObjectInputStreamTest {
    @Test
    void resolvesSerializedClassesWithTheProvidedLoader() throws Exception {
        byte[] encoded = SerializationUtils.serialize(new ArrayList<>(List.of("value")));

        try (
            ClassLoaderObjectInputStream input = new ClassLoaderObjectInputStream(
                new DenyingClassLoader(),
                new ByteArrayInputStream(encoded)
            )
        ) {
            assertThat(input.readObject()).isEqualTo(List.of("value"));
        }
    }

    @Test
    void resolvesSerializedProxyInterfacesWithTheProvidedLoader() throws Exception {
        Greeting greeting = Reflection.newProxy(Greeting.class, new GreetingHandler());
        byte[] encoded = SerializationUtils.serialize(greeting);

        try (
            ClassLoaderObjectInputStream input = new ClassLoaderObjectInputStream(
                new DelegatingClassLoader(),
                new ByteArrayInputStream(encoded)
            )
        ) {
            assertThat(((Greeting) input.readObject()).greet("proxy")).isEqualTo("Hello proxy");
        }
    }

    public static class DenyingClassLoader extends ClassLoader {
        public DenyingClassLoader() {
            super(null);
        }

        @Override
        public Class<?> loadClass(String name) throws ClassNotFoundException {
            throw new ClassNotFoundException(name);
        }
    }

    public static class DelegatingClassLoader extends ClassLoader {
        public DelegatingClassLoader() {
            super(ClassLoaderObjectInputStreamTest.class.getClassLoader());
        }
    }

    interface Greeting extends Serializable {
        String greet(String name);
    }

    public static class GreetingHandler implements InvocationHandler, Serializable {
        private static final long serialVersionUID = 1L;

        @Override
        public Object invoke(Object proxy, java.lang.reflect.Method method, Object[] arguments) {
            return "Hello " + arguments[0];
        }
    }
}
